---
abstract: "Individuals of European ancestry disproportionately dominate participation in human genetic studies, to the detriment of scientific inquiry and the equitable translation of genomic research. One genomic study type, expression quantitative trait locus (eQTL) mapping, identifies statistical associations between genetic variants and differential gene expression. However, whilst eQTL studies have been valuable tools for understanding the regulatory consequences of disease-associated genetic variants, not all eQTLs are shared across populations. Understanding the biological and genomic features of these eQTLs could enable prediction of eQTLs in understudied populations. 
Here we use summary statistics from three previous multi-population eQTL studies to classify eQTLs as population-specific or shared between at least two populations (African American, European American, Indonesian etc.). We train machine learning models to predict whether or not an eQTL is specific to its discovery population using publicly available information on the evolutionary, functional, and expression properties of these eQTLs (accuracy > 80%). 
Of all considered features, we find gene conservation (e.g. LOEUF, phyloP), Gene Ontology annotations, allele frequency and eQTL effect size are the most informative.This result suggests these features could be used to assess the probability a particular eQTL is specific to a population, and to understand gene regulatory trait differences across populations. Since current Eurocentric biases in genomic resources are likely to persist for some time, our approach could be an important step toward a more equitable understanding of gene regulation, and hence more equitable personalised medicine."
authors:
- Isobel Beasley
- Christina B Azodi
- Irene Gallego Romero 
date: "2021-11-17"
doi: ""
featured: true
image:
  caption: 'Image credit: [**Unsplash**](https://unsplash.com/photos/pLCdAaMFLTE)'
  focal_point: ""
  preview_only: false
links:
- name: 
  url: 
projects:
- internal-project
publication: In *Source Themes Conference*
publication_short: In *STC*
publication_types:
- "1"
publishDate: "2017-01-01T00:00:00Z"
slides: example
summary: Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis posuere tellus
  ac convallis placerat. Proin tincidunt magna sed ex sollicitudin condimentum.
tags:
- Source Themes
title: "Predicting the cross-population portability of human expression quantitative trait loci (eQTLs)"
url_code: '#'
url_dataset: '#'
url_pdf: 
url_poster: '#'
url_project: ""
url_slides: ""
url_source: '#'
url_video: '#'
---

{{% callout note %}}
Click the *Cite* button above to demo the feature to enable visitors to import publication metadata into their reference management software.
{{% /callout %}}

{{% callout note %}}
Create your slides in Markdown - click the *Slides* button to check out the example.
{{% /callout %}}

Abstract: 

Individuals of European ancestry disproportionately dominate participation in human genetic studies, to the detriment of scientific inquiry and the equitable translation of genomic research. One genomic study type, expression quantitative trait locus (eQTL) mapping, identifies statistical associations between genetic variants and differential gene expression. However, whilst eQTL studies have been valuable tools for understanding the regulatory consequences of disease-associated genetic variants, not all eQTLs are shared across populations. Understanding the biological and genomic features of these eQTLs could enable prediction of eQTLs in understudied populations. 

Here we use summary statistics from three previous multi-population eQTL studies to classify eQTLs as population-specific or shared between at least two populations (African American, European American, Indonesian etc.). We train machine learning models to predict whether or not an eQTL is specific to its discovery population using publicly available information on the evolutionary, functional, and expression properties of these eQTLs (accuracy > 80%). 

Of all considered features, we find gene conservation (e.g. LOEUF, phyloP), Gene Ontology annotations, allele frequency and eQTL effect size are the most informative.This result suggests these features could be used to assess the probability a particular eQTL is specific to a population, and to understand gene regulatory trait differences across populations. Since current Eurocentric biases in genomic resources are likely to persist for some time, our approach could be an important step toward a more equitable understanding of gene regulation, and hence more equitable personalised medicine.
